controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -86,25 +85,49 @@ public String getScopeName() {
      * @return : future of whether it was done or not
      */
     @Override
-    public CompletableFuture<Boolean> create(final StreamConfiguration configuration, long createTimestamp) {
-        final Create create = new Create(createTimestamp, configuration);
+    public CompletableFuture<CreateStreamResponse> create(final StreamConfiguration configuration, long createTimestamp) {
 
         return checkScopeExists()
-                .thenCompose(x -> checkStreamExists(create))
-                .thenCompose(x -> storeCreationTime(create))
-                .thenCompose(x -> createConfiguration(create))
-                .thenCompose(x -> createState(State.CREATING))
-                .thenCompose(x -> createSegmentTable(create))
-                .thenCompose(x -> createNewEpoch(0))
-                .thenCompose(x -> {
-                    final int numSegments = create.getConfiguration().getScalingPolicy().getMinNumSegments();
-                    final byte[] historyTable = TableHelper.createHistoryTable(create.getCreationTime(),
-                            IntStream.range(0, numSegments).boxed().collect(Collectors.toList()));
+                .thenCompose(x -> checkStreamExists(configuration, createTimestamp))
+                .thenCompose(createStreamResponse -> storeCreationTimeIfNotExists(createStreamResponse.getTimestamp())
+                        .thenCompose(x -> createConfigurationIfNotExists(createStreamResponse.getConfiguration()))
+                        .thenCompose(x -> createStateIfNotExists(State.CREATING))
+                        .thenCompose(x -> createNewSegmentTable(createStreamResponse.getConfiguration(), createStreamResponse.getTimestamp()))
+                        .thenCompose(x -> getState().thenCompose(state -> {
+                            if (state.equals(State.CREATING)) {
+                                return createNewEpoch(0);
```

Comment:
To clarify, this epoch zero is only created in the case the stream state is `CREATING` and we are assuming that as such there can't be any created epoch greater than zero. Given that epoch creation is idempotent, re-creating epoch zero for the stream in the case it happened before as part of the partial execution of create is not a problem. Is my reasoning correct?